### PR TITLE
Fix nodefile/nodelist

### PR DESCRIPTION
**Pull request by AymenFJA**
This fix is related to #2488.

### CODE REVIEW

File: src/radical/pilot/agent/launch_method/srun.py [hunk 1]
```diff
@@ -152,12 +161,11 @@ def get_launch_cmds(self, task, exec_path):
...
-        if self._vmajor > 18:
-            if nodefile:
-                mapping += ' --nodefile=%s' % nodefile
-        else:
-            if nodelist:
-                mapping += ' --nodelist=%s' % ','.join(str(n) for n in nodelist)
+        if nodefile:
+            mapping += ' --nodefile=%s' % nodefile
+
+        if nodelist:
```
Comment (andre-merzky) [inline src/radical/pilot/agent/launch_method/srun.py:167]: `elif`?  We should never have both.  In fact, I would say it is 
```
if nodefile:
    ...
else:
    ...
```

File: src/radical/pilot/agent/launch_method/srun.py [hunk 2]
```diff
@@ -129,15 +129,24 @@ def get_launch_cmds(self, task, exec_path):
...
+                if n_nodes > 42:
+                    nodefile = '%s/%s.nodes' % (sbox, uid)
+                    with ru.ru_open(nodefile, 'w') as fout:
+                        fout.write(','.join(nodelist))
+                        fout.write('\n')
+                    nodelist.clear()
+
+            # corener case that we can not address, related ticket:
+            # https://github.com/radical-cybertools/radical.pilot/issues/2488
+            elif self._vmajor <= 18 and n_nodes > 42:
```
Comment (andre-merzky) [inline src/radical/pilot/agent/launch_method/srun.py:147]: The check for `self._vmajor <= 18` seems redundant - in this else it's always true.  But also, we don't need to raise - we can the LM try with that long nodelist as command line arg anyway: either it works, hurray, or it fails, then we are not worse off than raising now...

**Review by andre-merzky — COMMENTED** [review]
Left 2 comments - thanks!

File: src/radical/pilot/agent/launch_method/srun.py [hunk 3]
```diff
@@ -129,15 +129,24 @@ def get_launch_cmds(self, task, exec_path):
...
-                nodefile = '%s/%s.nodes' % (sbox, uid)
-                with ru.ru_open(nodefile, 'w') as fout:
-                    fout.write(','.join(nodelist))
-                    fout.write('\n')
+                if n_nodes > 42:
+                    nodefile = '%s/%s.nodes' % (sbox, uid)
+                    with ru.ru_open(nodefile, 'w') as fout:
+                        fout.write(','.join(nodelist))
+                        fout.write('\n')
+                    nodelist.clear()
```
Comment (mtitov) [inline src/radical/pilot/agent/launch_method/srun.py:143]: this operation is unnecessary, latter you just check
```python
if nodefile:
    ...
elif nodelist:
    ...
```

File: src/radical/pilot/agent/launch_method/srun.py [hunk 4]
```diff
@@ -129,15 +129,24 @@ def get_launch_cmds(self, task, exec_path):
...
 
             # older slurm versions don't accept nodefiles
+            # 42 node is the upper limit to switch from `--nodelist`
+            # to `--nodefile`
             if self._vmajor > 18:
-                nodefile = '%s/%s.nodes' % (sbox, uid)
-                with ru.ru_open(nodefile, 'w') as fout:
-                    fout.write(','.join(nodelist))
-                    fout.write('\n')
+                if n_nodes > 42:
```
Comment (mtitov) [inline src/radical/pilot/agent/launch_method/srun.py:138]:
```python
if node > 42:

    if self._vmajor <= 18:
        raise Exception(...)

    nodefile = ...
```

btw I also would propose to move these numbers up, as constants (names for constants just an example, you can come up with yours)

```python
MIN_NUM_NODES_IN_NODELIST = 42
MIN_SLURM_VER_TO_NODELIST = 18
```

**Review by mtitov — COMMENTED** [review]
couple general comments

File: src/radical/pilot/agent/launch_method/srun.py (hunk 1 above)
Comment (AymenFJA) [reply src/radical/pilot/agent/launch_method/srun.py:167]: We would never have both @andre-merzky , as you can see if we have `nodfile` then once we finish writing to that `nodefile` from the `nodelist` we clear the `nodelist`. 

```python
with ru.ru_open(nodefile, 'w') as four:
        fout.write(','.join(nodelist))
        fout.write('\n')
nodelist.clear()

File: src/radical/pilot/agent/launch_method/srun.py (hunk 4 above)
Comment (AymenFJA) [reply src/radical/pilot/agent/launch_method/srun.py:138]: @mtitov  This looks good to set the min version of `slurm` and `number of nodes`

**Comment by AymenFJA** [comment]
@andre-merzky @mtitov thanks for your comments, I addressed them in the last commit, if this version is fine, then I can move forward with fixing the tests.

**Review by andre-merzky — APPROVED** [review]
LGTM, thanks!

**Review by mtitov — APPROVED** [review]
LGTM!

**Comment by andre-merzky** [comment]
A test got stuck in `pip install`, thus merged manually...

**Pull request merged**